JavaScript/chapter01/1.8 - Zero Matrix/rroque98_sol.js

```diff
@@ -0,0 +1,55 @@
+/*Write an algorithm such that if an element
+ in an M*N matrix is 0,it's entire row and column
+  are set to 0*/
+
+const zeroMatrix = nestedArr => {
+  const rowsAndColsWithZeros = checkForZeroIndex(nestedArr);
+  const rowLength = nestedArr[0].length;
+  const columnLength = nestedArr.length;
+  for (const rowIndex of rowsAndColsWithZeros['rows']) {
+    for (let x = 0; x < rowLength; x++) {
+      nestedArr[rowIndex][x] = 0;
+    }
+  }
+  for (const colIndex of rowsAndColsWithZeros['columns']) {
+    for (let x = 0; x < columnLength; x++) {
+      nestedArr[x][colIndex] = 0;
+    }
+  }
+  return nestedArr;
+
+  // **** Helper functions ****
+  function checkForZeroIndex(nestArr) {
+    let zeroIndices = { rows: [], columns: [] };
+    for (let i = 0; i < nestArr.length; i++) {
+      for (let x = 0; x < nestArr[i].length; x++) {
+        if (nestArr[i][x] === 0) {
+          zeroIndices['rows'].push(i);
```

Comment:
what would the runtime complexity of a matrix full of zeroes be?

Reply:
If we define N as the number of rows and M as the number of columns in the matrix, then I believe the time complexity would be O(NM)

Reply:
it is after the update, but it used to be O(max(N²M, NM²)), since the `zeroIndices` matrix would have up to O(max(N, M)) zeroes for each row and column, and each one of those would cause up to O(max(N, M)) updates per row/column.

now you're just counting each row/column just once.

Reply:
I see. I noticed I was double counting rows and columns that were already in `zeroIndices` after I made the helper function free and tried testing it for the desired behavior. Thanks for catching this and for the explanation.